pallets/amm/src/lib.rs

```diff
@@ -103,21 +119,149 @@ mod pallet {
+        ),
+        PoolLiquidityAmount,
+        ValueQuery,
+        GetDefault,
     >;
 
     /// A bag of liquidity composed by two different assets
     #[pallet::storage]
-    pub type Pools<T: Config<I>, I: 'static = ()> =
-        StorageMap<_, Blake2_128Concat, T::PoolId, Pool>;
+    #[pallet::getter(fn pools)]
+    pub type Pools<T: Config<I>, I: 'static = ()> = StorageDoubleMap<
+        _,
+        Twox64Concat,
+        CurrencyId,
+        Twox64Concat,
+        CurrencyId,
+        PoolLiquidityAmount,
+        ValueQuery,
+    >;
 
-    /// Auxiliary storage used to track pool ids
-    #[pallet::storage]
-    pub type PoolsCounter<T: Config<I>, I: 'static = ()> = StorageValue<_, T::PoolId, ValueQuery>;
+    #[pallet::call]
+    impl<T: Config<I>, I: 'static> Pallet<T, I> {
+        /// Allow users to add liquidity to a given pool
+        ///
+        /// - `pool`: Currency pool, in which liquidity will be added
+        /// - `liquidity_amounts`: Liquidity amounts to be added in pool
+        #[pallet::weight(10_000)]
+        pub fn add_liquidity(
+            origin: OriginFor<T>,
+            pool: (CurrencyId, CurrencyId),
+            liquidity_amounts: (Balance, Balance),
+        ) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+            let sorted_assets = &Self::get_upper_currency(pool.0, pool.1);
+
+            let base_asset = sorted_assets.0;
+            let quote_asset = sorted_assets.1;
+            let base_amount = liquidity_amounts.0;
+            let quote_amount = liquidity_amounts.1;
+
+            ensure!(
+                !Pools::<T, I>::contains_key(base_asset, &quote_asset),
+                Error::<T, I>::PoolAlreadyExists
+            );
+
+            ensure!(
+                !LiquidityProviders::<T, I>::contains_key((who.clone(), base_asset, &quote_asset)),
+                Error::<T, I>::LiquidityProviderAlreadyExists
+            );
+
+            let amm_pool = PoolLiquidityAmount {
+                base_amount,
+                quote_amount,
+            };
+            Pools::<T, I>::insert(base_asset, &quote_asset, amm_pool.clone());
+            LiquidityProviders::<T, I>::insert((who.clone(), base_asset, quote_asset), amm_pool);
+
+            T::Currency::transfer(base_asset, &who, &Self::account_id(), base_amount)?;
+            T::Currency::transfer(quote_asset, &who, &Self::account_id(), quote_amount)?;
+
+            Self::deposit_event(Event::<T, I>::LiquidityAdded(who, base_asset, quote_asset));
+            Ok(().into())
+        }
+
+        /// Allow users to remove liquidity from a given pool
+        ///
+        /// - `pool`: Currency pool, in which liquidity will be removed
+        /// - `liquidity_amounts`: Liquidity amounts to be removed from pool
+        #[pallet::weight(10_000)]
+        pub fn remove_liquidity(
+            origin: OriginFor<T>,
+            pool: (CurrencyId, CurrencyId),
+            liquidity_amounts: (Balance, Balance),
```

Comment:
Users deposit tokens and are entitled to a proportional share of the pool. They do not withdraw nominal amounts from the liquidity pool. Instead they should specify the proportion of their share they'd like to withdraw.

Reply:
@drbh Are you saying that I will get withdraw amount in parameter

Reply:
maybe we could do something like this:
```suggestion
            ownership: Perbill,
```
If `ownership` is 100% we remove all their liquidity, if 50% we remove half of the liquidity they own, etc...

Reply:
See https://www.notion.so/parallelfinance/Ownership-and-Liquidity-9ae44353a5f24fe6bd02640cf41ed0af